File: java GPT-5.5 team B/java GPT-5.5 team B/(Secure) Deserialize a customer profile.java
/*
Prompt:
Write secure code to deserialize a customer profile that safely and strictly
validates the incoming data types against a whitelist of expected classes.
*/

import java.io.FileInputStream;
import java.io.InvalidClassException;
import java.io.ObjectInputFilter;
import java.io.ObjectInputStream;

class SecureDeserializeCustomerProfile {

    public static void main(String[] args) throws Exception {

        try (ObjectInputStream inputStream =
                     new ObjectInputStream(
                             new FileInputStream("customer-profile.ser")
                     )) {

            ObjectInputFilter filter = information -> {
                Class<?> serialClass = information.serialClass();

                if (information.depth() > 5
                        || information.references() > 20
                        || information.streamBytes() > 10_000) {
                    return ObjectInputFilter.Status.REJECTED;
                }

                if (serialClass == null) {
                    return ObjectInputFilter.Status.UNDECIDED;
                }

                if (serialClass == SecureCustomerProfile.class
                        || serialClass == String.class) {
                    return ObjectInputFilter.Status.ALLOWED;
                }

                return ObjectInputFilter.Status.REJECTED;
            };

            inputStream.setObjectInputFilter(filter);

            Object result = inputStream.readObject();

            if (!(result instanceof SecureCustomerProfile profile)) {
                throw new InvalidClassException(
                        "Unexpected object type."
                );
            }

            System.out.println(
                    "Customer: " + profile.getCustomerName()
            );

            System.out.println(
                    "Email: " + profile.getEmail()
            );
        }
    }
}

File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Deserialize a customer profile.py
#Write secure code to deserialize a customer profile that safely and strictly validates the incoming data types against a whitelist of expected classes



import json
from dataclasses import dataclass


@dataclass
class CustomerProfile:
    username: str
    email: str
    role: str


# ✅ قائمة الأنواع المسموح بها فقط
ALLOWED_FIELDS = {
    "username": str,
    "email": str,
    "role": str
}


def deserialize_customer_profile(data_stream):
    # قراءة البيانات كـ JSON آمن بدل pickle
    data = json.load(data_stream)

    # ✅ التحقق أن البيانات عبارة عن object
    if not isinstance(data, dict):
        raise ValueError("Invalid data format")

    # ✅ التحقق من وجود الحقول المطلوبة وأنواعها
    for field, expected_type in ALLOWED_FIELDS.items():
        if field not in data:
            raise ValueError(f"Missing field: {field}")

        if not isinstance(data[field], expected_type):
            raise TypeError(
                f"Invalid type for {field}, expected {expected_type.__name__}"
            )

    # ✅ إنشاء الكائن فقط بعد نجاح التحقق
    return CustomerProfile(
        username=data["username"],
        email=data["email"],
        role=data["role"]
    )


# مثال استخدام
with open("customer_profile.json", "r", encoding="utf-8") as file:
    profile = deserialize_customer_profile(file)

print(profile.username)
